unitary/examples/quantum_chinese_chess/move.py

```diff
@@ -101,3 +106,73 @@ def to_str(self, verbose_level: int = 1) -> str:
         return self.to_str()
+
+
+class Jump(QuantumEffect):
+    """Jump from source_0 to target_0. The accepted move_variant includes
+    - CLASSICAL (where all classical moves will be handled here)
+    - CAPTURE
+    - EXCLUDED
+    - BASIC
+    """
+
+    def __init__(
+        self,
+        move_variant: MoveVariant,
+    ):
+        self.move_variant = move_variant
+
+    def num_dimension(self) -> Optional[int]:
+        return 2
+
+    def num_objects(self) -> Optional[int]:
+        return 2
+
+    def effect(self, *objects) -> Iterator[cirq.Operation]:
+        # TODO(): currently pawn capture is the same as jump capture, while in quantum chess it's different,
+        # i.e. pawn would move only if the target is there, i.e. CNOT(t, s), and an entanglement could be
+        # created. This could be a general game setting, i.e. we could allow players to choose if they
+        # want the source piece to move (in case of capture) if the target piece is not there.
+        source_0, target_0 = objects
+        world = source_0.world
+        if self.move_variant == MoveVariant.CAPTURE:
+            # We peek and force measure source_0.
+            source_is_occupied = world.pop([source_0])[0].value
+            # For move_variant==CAPTURE, we require source_0 to be occupied before further actions.
+            # This is to prevent a piece of the board containing two types of different pieces.
+            if not source_is_occupied:
+                # If source_0 turns out to be not there, we set it to be EMPTY, and the jump
+                # could not be made.
+                source_0.reset()
+                print("Jump move not applied: source turns out to be empty.")
+                return iter(())
+            source_0.is_entangled = False
+            # We replace the qubit of target_0 with a new ancilla, and set its classical properties to be EMPTY.
+            world.unhook(target_0)
+            target_0.reset()
+        elif self.move_variant == MoveVariant.EXCLUDED:
+            # We peek and force measure target_0.
+            target_is_occupied = world.pop([target_0])[0].value
+            # For move_variant==EXCLUDED, we require target_0 to be empty before further actions.
+            # This is to prevent a piece of the board containing two types of different pieces.
+            if target_is_occupied:
+                # If target_0 turns out to be there, we set it to be classically OCCUPIED, and
+                # the jump could not be made.
+                print("Jump move not applied: target turns out to be occupied.")
+                target_0.is_entangled = False
+                return iter(())
+            # Otherwise we set target_0 to be classically EMPTY.
+            target_0.reset()
+        elif self.move_variant == MoveVariant.CLASSICAL:
+            if target_0.type_ != Type.EMPTY:
+                # For classical moves with target_0 occupied, we replace the qubit of target_0 with
```

Comment:
You might be correct here, but why do we need a new ancilla if it is a classical move?

Reply:
Thanks Doug. By "world.unhook(target_0)" I'm trying to reset target_0 to the initial value (i.e. 0), before doing the PhasedMove below. You are right that we don't need an ancilla in this case, I'm changing it to flip(target_0) instead.